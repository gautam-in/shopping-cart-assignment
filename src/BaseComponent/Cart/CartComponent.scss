/********************* CartComponent ************************/
@import "../../Styles/color-variable.scss";
@import "../../Styles/variable.scss";
@import "../../Styles/mixin.scss";

.cart {
  display: grid;
  grid-template-columns: 1fr 2fr;
  background-color: $LIGHT_GREY;
  gap: $SMALL_GAP;
  text-decoration: none;

  & .cart_icon {
    width: $MEDIUM_WIDTH;
    height: auto;
    justify-self: end;
    align-self: center;
  }

  & .cart_items {
    font-size: $MEDIUM_FONT_SIZE;
    color: $APP_BLACK;
    justify-self: start;
    align-self: center;
  }
}
