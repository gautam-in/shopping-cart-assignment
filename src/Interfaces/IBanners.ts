export default interface IBanners {
    id: string;
    order: number;
    isActive: boolean;
    bannerImageUrl: string;
    bannerImageAlt: string;
}
