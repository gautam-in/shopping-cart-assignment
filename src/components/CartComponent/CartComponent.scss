.cart {
    background-color: #ece9e9;
    text-align: center;
    //margin-bottom: 40px;
    .heading {
      font-size: large;
      padding: 10px 0px 0px 10px;
      margin-bottom: 30px solid #ece9e9;
      text-align: start;
      background-color: white;
      box-sizing: border-box;
    }
    .empty-cart {
      // display: flex;
      text-align: center;
      padding: 140px 0px;
      div {
        justify-content: center;
        align-items: center;
        text-align: center;
        padding: 20px;
        padding-bottom: 0px;
        font-weight: 600;
        font-family: "Dosis", sans-serif;
      }
  
      p {
        //width: 200px;
        margin: 0px;
        //padding-left: 100px;
        justify-content: center;
        align-items: center;
        text-align: center;
      }
    }
  
    .checkout {
      margin-top: 150px;
      background-color: #c91e46;
  
      .checkout-btn {
        display: flex;
        width: inherit;
        margin: 0px 40px;
        span {
          padding-left: 750px;
        }
      }
    }
  }
  
  @media (min-width: 769px) {
    .modal-header {
      background-color: black;
      color: white;
      font-family: "Dosis", sans-serif;
    }
    .modal-body {
      padding-top: 0px;
      padding-left: 50px;
    }
    .checkout {
      max-width: 465px;
      max-height: 80px;
  
      .proceed-btn {
        max-height: 50px;
        display: flex;
      }
      .checkout-btn {
        max-width: 500px;
  
        span {
          padding-left: 170px !important;
          max-width: 300px !important;
          max-height: 50px;
          text-align: end;
        }
      }
    }
  }