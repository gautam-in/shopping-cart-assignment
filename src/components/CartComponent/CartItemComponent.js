import ButtonComponent from "../../components/CustomComponent/ButtonComponent/ButtonComponent";
import { useDispatch } from "react-redux";
import { sendAddRequest } from "../../store/cart-actions";
import { cartActions } from "../../store/cartSlice";
import "./CartItemComponent.scss";
const CartItemComponent = ({ product }) => {
  const dispatch = useDispatch();
  //console.log(product);
  const addToCartHandler = (product) => {
    dispatch(sendAddRequest(product));
  };
  const decreaseItemQty = (product) => {
    dispatch(cartActions.reduceQuantity({ id: product.id }));
  };
  return (
    <div className="cart-item-row">
      <img className="cart-img" src={product.imageUrl} alt={product.name} />
      <div className="product-details">
        <div>
          <div className="product-name">{product.name}</div>

          <div className="d-flex">
            <ButtonComponent
              className="cart-butn"
              onClick={() => decreaseItemQty(product)}
            >
              -
            </ButtonComponent>
            {product.quantity}
            <ButtonComponent
              className="cart-butn"
              onClick={() => addToCartHandler(product)}
            >
              +
            </ButtonComponent>{" "}
            x Rs.{product.price}
          </div>
        </div>
        <div className="item-total-price">Rs.{product.totalItemPrice}</div>
      </div>
    </div>
  );
};

export default CartItemComponent;