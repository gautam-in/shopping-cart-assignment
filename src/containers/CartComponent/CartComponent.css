.CartWrapper {
  height: 80vh;
  background-color: #e8eaeb;
}
.CartHeader {
  background-color: #16191b;
  display: flex;
  justify-content: space-between;
  color: #feffff;
  padding: 0 1rem 0 2rem;
}
.CloseIcon {
  cursor: pointer;
  height: 2rem;
  width: auto;
  filter: invert(1);
}
.CartHeaderImg {
  display: flex;
  justify-content: center;
  align-items: center;
}
.CartContent {
  color: #1a1b1c;
  height: 64vh;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}
.CartCardContent {
  justify-content: unset;
  align-items: unset;
  overflow: hidden;
  overflow-y: scroll;
  height: 61vh;
}
.CustomBtnTab {
  width: 96%;
  display: flex;
  justify-content: space-between;
  font-weight: normal;
}
.CustomBtnTabNoItem {
  width: 96%;
  font-weight: normal;
}
h3 {
  margin: 0;
}

.marLR {
  margin: 0 10px;
}

.marL {
  margin-left: 10px;
}

.CartFooter {
  position: fixed;
  bottom: 0;
  width: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  background-color: white;
}
.FooterInfo {
  margin-top: 10px;
}
.CartContent .CartCardContent :last-child {
  margin-bottom: 1rem;
}
.LowestPriceImg {
  width: 100%;
  height: auto;
}
.LowestPriceImgWrap {
  width: 7rem;
  padding: 10px;
}
.LowestPriceWrap {
  display: flex;
  align-items: center;
  margin: 10px;
  margin-bottom: 1rem;
  background-color: white;
}
@media (max-width: 499px) {
  .CustomBtnTab,
  .CustomBtnTabNoItem {
    font-size: 15px;
  }
}

@media (min-width: 500px) and (max-width: 1024px) {
  .CartCardContent {
    height: 65vh;
  }
  .CartContent {
    height: 65vh;
  }
}
