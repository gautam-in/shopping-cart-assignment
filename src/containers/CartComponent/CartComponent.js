import React from "react";
import { connect } from "react-redux";
import Button from '../../components/UI/Button/Button'
import CartCard from '../../components/UI/CartCard/CartCard'
import * as actions from "../../store/actions/index";

import classes from "./CartComponent.css";

const CartComponent = (props) => {
    const modCartList = {};
    const newCartList = [...props.cartList];
    newCartList.sort((a, b) =>
        a.name > b.name ? 1 : b.name > a.name ? -1 : 0
    );
    console.log(newCartList);
    let totalPrice = 0;
    newCartList && newCartList.length > 0 && newCartList.forEach(item => {
        if (modCartList[item.id]) {
            modCartList[item.id].push(item);
        } else {
            modCartList[item.id] = [item];
        }
        totalPrice += item.price;
    })
    console.log("count", modCartList);
    let cartContent = null;
    if (newCartList && newCartList.length === 0) {
        cartContent = <div>
            <div className={classes.CartContent}>
                <h3>No items in your cart</h3>
                <p>Your favourite items are just a click away</p>
            </div>
            <Button CustomBtn={classes.CustomBtnTabNoItem}
                clicked={() => props.onCartClick(false)}>Start Shopping</Button>
        </div>
    } else {
        cartContent = <div>
            <div className={[classes.CartContent, classes.CartCardContent].join(' ')}>
                {Object.keys(modCartList).length > 0 &&
                    Object.keys(modCartList).map(id => <CartCard data={modCartList[id][0]} count={modCartList[id].length} key={id} addToCartHandler={props.onAddToCart} />)
                }
                <div className={classes.LowestPriceWrap}>
                    <div className={classes.LowestPriceImgWrap}>
                        <img className={classes.LowestPriceImg} src={`../../../static/images/lowest-price.png`} alt="Lowest-price img" />
                    </div>
                    {"You won't find it cheaper anywhere"}
                </div>

            </div>
            <div className={classes.CartFooter}>
                <div className={classes.FooterInfo}>Promocode can be applied on payment page</div>
                <Button CustomBtn={classes.CustomBtnTab}
                    clicked={() => props.onCartClick(false)}>
                    <div className={classes.marLR}>Proceed to Checkout</div>
                    <div className={classes.marLR}>
                        <span >{`Rs. ${totalPrice}`}</span>
                        <span className={classes.marL}>{">"}</span>
                    </div>
                </Button>
            </div>

        </div >
    }

    return (
        <div className={classes.CartWrapper}>
            <div className={classes.CartHeader}>
                <h4>My Cart({newCartList.length} item)</h4>
                <div className={classes.CartHeaderImg}>
                    <img src="../../../static/images/close.svg" alt="Close img" onClick={() => props.onCartClick(false)} className={classes.CloseIcon} /></div>
            </div>
            {cartContent}
        </div>
    )
};

const mapStateToProps = (state) => {
    return {
        cartList: state.product.addToCartProducts,
        cartFlag: state.product.cartFlag
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        onCartClick: () => dispatch(actions.cartHandler()),
        onAddToCart: (product, type) => dispatch(actions.addToCart(product, type)),
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(CartComponent);
