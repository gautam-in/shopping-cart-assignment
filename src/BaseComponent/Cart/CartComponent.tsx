import "./CartComponent.scss";
import { ImagesConstantIndex } from "../../Utility/Images/ImagesConstantIndex";
import { Link } from "react-router-dom";
import { ROUTE_URL } from "../../Utility/RouteUrlConstant";
import { IState } from "../../Store/Store";
import { connect } from "react-redux";
import { getCartItemCount } from "../../Store/Selectors";

interface ICartProps {
    className?: string;
}

interface IStoreProps {
    cartItemCount: number;
}

type TCartComponentProps = IStoreProps & ICartProps;

const CartComponent: React.FC<TCartComponentProps> = (
    props: TCartComponentProps
) => {
    const BASE_CLASSNAME: string = "cart";
    const CART_ICON_CLASSNAME: string = `${BASE_CLASSNAME}_icon`;
    const CART_ITEMS_CLASSNAME: string = `${BASE_CLASSNAME}_items`;
    const CLASSES: string = `${BASE_CLASSNAME} ${props.className}`;
    return (
        <Link to={ROUTE_URL.CART_PAGE} className={CLASSES}>
            <img
                src={ImagesConstantIndex.CART}
                alt="cart"
                className={CART_ICON_CLASSNAME}
            />
            <p className={CART_ITEMS_CLASSNAME}>{props.cartItemCount} items</p>
        </Link>
    );
};

function mapStateToProps(state: IState): IStoreProps {
    return {
        cartItemCount: getCartItemCount(state.appState),
    };
}

export default connect<IStoreProps, {}, {}, IState>(mapStateToProps)(
    CartComponent
);
