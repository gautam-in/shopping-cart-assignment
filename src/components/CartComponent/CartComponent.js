import { useDispatch, useSelector } from "react-redux";
import { useNavigate } from "react-router-dom";
import ButtonComponent from "../../components/CustomComponent/ButtonComponent/ButtonComponent";
import CartItemComponent from "./CartItemComponent";
import "./CartComponent.scss";

import { uiActions } from "../../store/uiSlice";
const CartComponent = () => {
  const navigation = useNavigate();
  const dispatch = useDispatch();
  const cartItems = useSelector((state) => state.cartReducer.cartItems);
  const cartTotalQty = useSelector((state) => state.cartReducer.totalQty);
  const cartTotalPrice = useSelector((state) => state.cartReducer.totalPrice);
  const isMobile = useSelector((state) => state.uiReducer.isMobile);
  const isTablet = useSelector((state) => state.uiReducer.isTablet);
  //console.log(cartItems)
  return (
    <div className="cart">
      {(isMobile || isTablet) && (
        <>
          <div style={{ height: "30px" }}></div>
          <div className=" heading l-text">
            My Cart{" "}
            {cartTotalQty
              ? cartTotalQty > 1
                ? `(${cartTotalQty} items)`
                : `(${cartTotalQty} item)`
              : ""}
          </div>
        </>
      )}
      <div style={{ height: "20px" }}></div>
      {cartItems.length > 0 && (
        <>
          {cartItems.map((item) => (
            <CartItemComponent key={item.id} product={item} />
          ))}
          <div style={{ height: "30px" }}></div>
          <div className="cart-item-row">
            <img
              src={`/static/images/lowest-price.png`}
              alt="guaranteed banner"
            />
            <div className=" text" style={{ paddingLeft: "20px" }}>
              You won't find it cheaper anywhere
            </div>
          </div>
          <div className="checkout text">
            <div style={{ backgroundColor: "white", textAlign: "center" }}>
              Promo code can be applied on payment
            </div>
            <ButtonComponent className="proceed-btn">
              <div className="checkout-btn">
                <div className="justify-content-start">Proceed to Checkout</div>
                <span>
                  Rs.{cartTotalPrice} {`>`}
                </span>
              </div>
            </ButtonComponent>
          </div>
        </>
      )}
      {!cartItems.length > 0 && (
        <>
          <div className="empty-cart">
            <div>No items in your cart.</div>
            <p className="text">Your favorite items are just a click away</p>
          </div>
          <ButtonComponent
            style={{ width: "300px" }}
            onClick={() => {
              navigation("/products");
              dispatch(uiActions.setshowCart({ show: false }));
            }}
          >
            Start Shopping
          </ButtonComponent>
        </>
      )}
    </div>
  );
};

export default CartComponent;