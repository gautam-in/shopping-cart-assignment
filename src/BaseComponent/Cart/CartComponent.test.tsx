import { ShallowWrapper, shallow } from "enzyme";
import IBanners from "../../Interfaces/IBanners";
import { CartComponent, TCartComponentProps } from "./CartComponent";

describe("CartComponent", () => {
    const cartItemCount = 0;
    const tree: ShallowWrapper<TCartComponentProps> = shallow(
        <CartComponent cartItemCount={cartItemCount} />
    );

    test("CartComponent Snapshot render", () => {
        expect(tree).toMatchSnapshot();
    });
});