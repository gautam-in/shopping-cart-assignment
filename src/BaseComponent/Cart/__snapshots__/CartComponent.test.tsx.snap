// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartComponent CartComponent Snapshot render 1`] = `ShallowWrapper {}`;
